File: ios/ReactNativeChat/GoogleAnalyticsBridge.m
//
//  GoogleAnalyticsBridge.m
//  ReactNativeChat
//
//  Created by Leland Richardson on 4/24/17.
//  Copyright © 2017 Facebook. All rights reserved.
//
#import "GoogleAnalyticsBridge.h"
#import <React/RCTConvert.h>
#import <Google/Analytics.h>


@implementation GoogleAnalyticsBridge {

}

RCT_EXPORT_MODULE();

RCT_EXPORT_METHOD(screenView:(NSString *)screenName)
{
  id<GAITracker> tracker = [[GAI sharedInstance] defaultTracker];
  [tracker set:kGAIScreenName value:screenName];
  [tracker send:[[GAIDictionaryBuilder createScreenView] build]];
}

RCT_EXPORT_METHOD(trackEvent:(NSString *)category action:(NSString *)action label:(NSString *)label value:(NSNumber *)value)
{
  id<GAITracker> tracker = [[GAI sharedInstance] defaultTracker];
  [tracker send:[[GAIDictionaryBuilder createEventWithCategory:category
                                                        action:action
                                                         label:label
                                                         value:value] build]];
}

RCT_EXPORT_METHOD(setUser:(NSString *)user)
{
  id<GAITracker> tracker = [[GAI sharedInstance] defaultTracker];
  [tracker set:kGAIUserId value:user];
}

@end

File: android/app/src/main/java/com/reactnativechat/GoogleAnalyticsBridge.java
package com.reactnativechat;

import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.google.android.gms.analytics.GoogleAnalytics;
import com.google.android.gms.analytics.HitBuilders;
import com.google.android.gms.analytics.Tracker;

public class GoogleAnalyticsBridge extends ReactContextBaseJavaModule {

  private Tracker mTracker;

  public GoogleAnalyticsBridge(ReactApplicationContext reactContext) {
    super(reactContext);
  }

  synchronized public Tracker getDefaultTracker() {
    if (mTracker == null) {
      GoogleAnalytics analytics = GoogleAnalytics.getInstance(getReactApplicationContext());
      // To enable debug logging use: adb shell setprop log.tag.GAv4 DEBUG
      mTracker = analytics.newTracker(R.xml.global_tracker);
    }
    return mTracker;
  }

  @Override
  public String getName() {
    return "GoogleAnalyticsBridge";
  }

  synchronized GoogleAnalytics getAnalyticsInstance() {
    return GoogleAnalytics.getInstance(getReactApplicationContext());
  }

  @ReactMethod
  public void screenView(String screenName) {
    Tracker tracker = getDefaultTracker();
    tracker.setScreenName(screenName);
    tracker.send(new HitBuilders.ScreenViewBuilder().build());
  }

  @ReactMethod
  public void trackEvent(String category, String action, String label, Integer value) {
    Tracker tracker = getDefaultTracker();
    HitBuilders.EventBuilder builder = new HitBuilders.EventBuilder();

    builder.setCategory(category);
    builder.setAction(action);

    // if (optionalValues.hasKey("label")) {
    //   builder.setLabel(optionalValues.getString("label"));
    // }
    // if (optionalValues.hasKey("value")) {
    //   builder.setValue(optionalValues.getInt("value"));
    // }

    tracker.send(builder.build());
  }

  @ReactMethod
  public void setUser(String userId) {
    Tracker tracker = getDefaultTracker();
    tracker.set("&uid", userId);
  }

}
